qualtran/bloqs/phase_estimation/lp_resource_state.py

```diff
@@ -0,0 +1,179 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""Resource states proposed by A. Luis and J. Peřina (1996) for optimal phase measurements"""
+from functools import cached_property
+from typing import Set, TYPE_CHECKING
+
+import attrs
+import cirq
+import numpy as np
+from numpy._typing import NDArray
+
+from qualtran import bloq_example, BloqDocSpec, GateWithRegisters, QUInt, Register, Side, Signature
+from qualtran.bloqs.basic_gates import CZPowGate, Hadamard, OnEach, Ry, Rz, XGate, ZPowGate
+from qualtran.bloqs.mcmt import MultiControlPauli
+from qualtran.cirq_interop import CirqGateAsBloq
+from qualtran.cirq_interop.t_complexity_protocol import TComplexity
+
+if TYPE_CHECKING:
+    from qualtran.resource_counting import BloqCountT, SympySymbolAllocator
+
+
+@attrs.frozen
+class _LPRSHelper(GateWithRegisters):
+    """Helper Bloq to prepare an intermediate resource state which can be used in AA
+
+    Prepares the intermediate resource state described in Eq 19 of
+    https://arxiv.org/pdf/1805.03662.pdf, which can then be used in a single round of
+    Amplitude Amplification to boost the amplitude of desired resource state to 1.
+    """
+
+    bitsize: int
+    eps: float = 1e-11
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature.build(m=self.bitsize, anc=1)
+
+    def short_name(self) -> str:
+        return 'LPRS'
+
+    def decompose_from_registers(
```

Comment:
What's the general motivation for making this a GateWithRegisters rather than Bloq/build_composite_bloq? I guess below we're missing a global phase operations for bloqs right?

Reply:
Yes, also the controlled bloq in Qualtran isn't very functional right now for simulations; so we need to fallback to `cirq.ControlledGate`. See https://github.com/quantumlib/Qualtran/issues/650 for example